doc/source/dev/contributor/quickstart_docker.rst

```diff
@@ -0,0 +1,177 @@
+   ``scipy`` folder containing the SciPy source code. This assumes that
+   you have a ``git`` command line client that is available on your
+   PATH; if not, you can follow these `instructions to install a git client`_.
+
+Starting Docker
+---------------
+
+Instructions for getting started with Docker can be found `here`_. After
+ensuring that Docker is working correctly, follow the instructions below to
+start a Docker container for SciPy development. You'll follow the same
+instructions each time you want to start the container, as changes made to a
+container do not persist after you close it.
+
+#. In a terminal window, change the directory (using the ``cd`` command)
+   to the folder that contains the SciPy git repository.
+
+#. Ensure that Docker Desktop (or Docker Toolbox) is running, and start up the
+   SciPy Docker container by entering the following command in a terminal
+   window::
+
+      docker run -it --rm -v $PWD/scipy:/home/scipy scipy/scipy-dev /bin/bash
+
+   This command starts (``run``) an interactive (``-it``) Docker container
+   named ``scipy-dev`` (based on Ubuntu Bionic) from the ``scipy``
+   `Docker Hub repository`_. When the Docker container starts, the
+   ``scipy`` directory from the current directory of the host (``$PWD``) is
+   made available in the container as ``/home/scipy``. The changes you make
+   from the container to any of the files in that directory are also
+   visible in the host, and vice-versa.
+
+#. You should now be in the container, with something like::
+
+      root@468e1b9564e4:/#
+
+   as a prompt.
+
+#. Navigate to the SciPy source directory, which is shared with the host OS.
+
+   ::
+
+      cd /home/scipy
+
+#. The container has both Python 3.6 and Python 3.7 available. To start
+   using/building SciPy, we need to install some dependencies::
+
+      pip3.7 install numpy cython pytest pybind11
+
+   If you want to work with Python 3.6 use the ``pip3.6`` command instead.
+
+#. Do an in-place build by entering::
+
+      python3.7 setup.py build_ext --inplace
+
+   This will compile the C,
+   C++, and Fortran code that comes with SciPy. ``setup.py`` is a
+   script in the root directory of SciPy, which is why you have to be
+   in the SciPy root directory to call it. ``build_ext`` is a command
+   defined in ``setup.py``, and ``--inplace`` is an option we’ll use to
+   ensure that the compiling happens in the SciPy directory you already
+   have rather than some other folder on your computer. If you want to
+   work with Python 3.6, replace ``python3.7`` with ``python3.6``.
+
+#. Test the build by entering::
+
+      python3.7 runtests.py -v
+
+   ``runtests.py`` is another script in the SciPy root directory. It runs a
+   suite of tests that make sure SciPy is working as it should, and ``-v``
+   activates the ``–verbose`` option to show all the test output.
+
+From here, you can start a Python console (e.g. enter ``python3.7``) or
```

Comment:
Is this the best way of doing this? I normally do a `pip install .`, navigate to a different directory and start using Python.

Reply:
If you install that way, don't you have to install again after making changes to the source code?

Reply:
> Is what the best way?

Of having the changes you just made to scipy code available in a newly opened interpreter.

> Do you have to do pip install . every time you change the source code?

If I'm not using runtests.py that's what I currently do. But if there are easier ways I'm all for them.

Reply:
I think `pip install -e .` makes symlinks instead of copying files so you don't have to run it after making changes to source for changes to take effect.

Reply:
I haven't used that before, though. I should try it out.

Adding the folder to PYTHONPATH, `conda develop`, and `python setup.py develop` have all made it so I don't have to repeat the install, but adding to PYTHONPATH was the first I learned so I guess I stuck with that.

Reply:
@rgommers do you have a recommendation for which of the options we should add to these instructions?

- `pip install -e .`
- `python setup.py develop`
- manually add SciPy root to PYTHONPATH

Note that in the other quickstarts, `conda develop` is used. I think that is appropriate there because `conda` is used to install everything and manage virtual environments, but this tutorial doesn't use Anaconda.